lib/metanorma/collection_renderer.rb

```diff
@@ -0,0 +1,267 @@
+    # @param read [Boolean]
+    # @return [Array<String, nil>]
+    def ref_file(ref, read)
+      file = File.read(ref, encoding: "utf-8") if read
+      filename = ref.sub(/\.xml$/, ".html")
+      [file, filename]
+    end
+
+    # @param id [String]
+    # @param read [Boolean]
+    # @return [Array<String, nil>]
+    def xml_file(id, read)
+      file = @xml.at(ns("//doc-container[@id = '#{id}']")).to_xml if read
+      filename = id + ".html"
+      [file, filename]
+    end
+
+    def update_bibitem(b, docid, identifier)
+      docid = b&.at(ns("./docidentifier"))&.text
+      unless @files[docid]
+        warn "Cannot find crossreference to document #{docid} in document #{identifier}!"
+        abort
+      end
+      id = b["id"]
+      newbib = b.replace(@files[docid][:bibdata])
+      newbib.name = "bibitem"
+      newbib["id"] = id
+      newbib&.at(ns("./ext"))&.remove
+      file, url = targetfile(@files[docid], false)
+      newbib.at(ns("./docidentifier")).previous = %{<uri type="citation">#{url}</uri>}
+    end
+
+    def repo_docid(docid)
+      docid.sub(%r{^current-metanorma-collection/}, "")
+    end
+
+    # TODO: update crossreferences to other files in the selection
+    # repo(current-metanorma-collection/ISO 17301-1:2016)
+    # replaced by
+    # bibdata of "ISO 17301-1:2016" in situ as bibitem
+    # Any erefs to that bibitem id are replaced with relative URL
+    # Preferably with anchor, and is a job to realise dynamic lookup of localities
+    def update_xrefs(file, identifier)
+      docxml = Nokogiri::XML(file)
+      docxml.xpath(ns("//bibitem[not(ancestor::bibitem)]")).each do |b|
+        docid = b&.at(ns("./docidentifier[@type = 'repository']"))&.text
+        next unless docid && %r{^current-metanorma-collection/}.match(docid)
+
+        update_bibitem(b, docid, identifier)
+        update_anchors(b, docxml, docid)
+      end
+      docxml.to_xml
+    end
+
+    # if there is a crossref to another document, with no anchor, retrieve the
+    # anchor given the locality, and insert it into the crossref
+    def update_anchors(b, docxml, id)
+      docid = b&.at(ns("./docidentifier"))&.text
+      docxml.xpath("//xmlns:eref[@citeas = '#{docid}']").each do |e|
+        e.at(ns(".//locality[@type = 'anchor']")).nil? or next
+        ins = e.at(ns("./localityStack")) or next
+        type = ins&.at(ns("./locality/@type"))&.text
+        ref = ins&.at(ns("./locality/referenceFrom"))&.text
+        anchor = @files[docid][:anchors][type][ref] and
+          ins << %(<locality type="anchor"><referenceFrom>#{anchor.sub(/^_/, '')}</referenceFrom></locality>)
+      end
+    end
+
+    # process each file in the collection
+    # files are held in memory, and altered as postprocessing
+    def files
+      @files.each do |identifier, x|
+        file, filename = targetfile(x, true)
+        file = update_xrefs(file, identifier)
+        Tempfile.open(["collection", ".xml"], encoding: "utf-8") do |f|
```

Comment:
What if I already have a compiled set of child files? We should reuse those instead of re-compiling them?

Reply:
Do we suppose the compiled files in the folder the collection is?

Reply:
> Do we suppose the compiled files are in the folder the collection is?

Let's allow definition of the locations in a YAML file?

If the location is not defined they should be in an expected location.

Reply:
This command updates a file content:
```
file = update_xrefs(file, identifier)
```
so I think we can't reuse existed file.